README.org
* ebDps - DPS5020 - Control Digital Power Supply  ebench control $Release:0.0.6-SNAPSHOT$
:PROPERTIES:
:TOC:      :include all
:END:


:CONTENTS:
- [[#ebdps---dps5020---control-digital-power-supply--ebench-control-release004][ebDps - DPS5020 - Control Digital Power Supply  ebench control $Release:0.0.6-SNAPSHOT$]]
- [[#references][References]]
- [[#usage][Usage]]
  - [[#running-ebdps--tool][Running ebDps -tool]]
    - [[#command-line-use][Command line use]]
    - [[#interactive-usage][Interactive usage]]
  - [[#options][Options]]
  - [[#---getting-help][? - getting help]]
- [[#commands-and-parameters][Commands and parameters]]
  - [[#status---read-dsp-status][status - Read DSP status]]
  - [[#off---turn-off][off - Turn off]]
  - [[#set--configure][set -configure]]
  - [[#on-turn-on][on turn on]]
  - [[#csv---csv-driver][csv - CSV driver]]
  - [[#scan_ports---list-tty-devices][scan_ports - list tty devices]]
  - [[#_ttys---check-serial-device][_ttys - Check serial device]]
  - [[#_info--command][_info -command]]
- [[#document-information][Document information]]
  - [[#version-used][Version used]]
  - [[#documents][Documents]]
:END:


* References


Wrapper for ~dps_modbus.py~
https://forum-raspberrypi.de/forum/thread/46324-ansteuerung-programmierbarer-stepdown-regler-dps/
for controlling Digital Power Supply DPS8020/5020 in
https://www.mediafire.com/folder/g21va7a0v2c15/power-2#v6b16r86l6xxb


* Usage

** Running ~ebDps~ -tool

*** Command line use

 For command line use pass commands and paramters to command ~ebenchDps~

 #+BEGIN_SRC bash :eval no-export :results output :noweb yes :exports results
 ebDps ? | grep Usage
 #+END_SRC

 #+RESULTS:
 : Usage: ebDps [options] [commands and parameters] 


*** Interactive usage

 Running the ~ebDps~ tool without commands line parameters starts
 interactive session, and the tool prompts for commands and parameters

 #+BEGIN_SRC bash :eval no-export :results output :noweb yes :exports results
 cat << EOF | ebDps 
 q
 EOF
 #+END_SRC

 #+RESULTS:
 : [q=quit,?=commands,??=help on command] > 


** Options

Supported command line options are

#+BEGIN_SRC bash :eval no-export :results output :noweb yes :exports results
# ebenchDps/ebenchDps_main.py --help
ebDps --helpfull | sed -n '/ebenchDps_main/,/^$/p'
#+END_SRC

#+RESULTS:
: ebenchDps.ebenchDps_main:
:   --port: MODBUS device port
:     (default: '/dev/ttyUSB0')
: 

Refer command ~_ttys~ and ~scan_ports~ how to locate device port,
where DPS -device can be reached.


** ~?~ - getting help

Running

 #+name: listCommands
 #+BEGIN_SRC bash :eval no-export :results output :noweb yes :exports both
 ebDps ?
 #+END_SRC

list commands supported by the tool

 #+RESULTS: listCommands
 #+begin_example
 ebenchDps: Control DPS8020 Digital Power Supply

 Usage: ebDps [options] [commands and parameters] 

 Commands:

 ---------- ebDps  ----------
          status  : Read DSP status
             set  : Configure
              on  : Turn on
             csv  : CSV driver
             off  : Turn off
 ---------- Management ----------
      scan_ports  : Scan for device serial address
 ----------    Util    ----------
               !  : Start recording
               .  : Stop recording
          screen  : Take screenshot
               ?  : List commands
              ??  : List command parameters
 ----------    Quit    ----------
               q  : Exit


 The tool :

 - query Digital Power supply status e.g. voltage/current setpoint
 - configure voltage and current setpoint
 - turn output on and off
 - use CSV file to generate output voltage output


 #+end_example

To get more detailed help use ~??~.  For example, to running
run

#+name: help-set
#+BEGIN_SRC bash :eval no-export :results output :noweb yes :exports both
 ebDps ?? command=set
#+END_SRC

gives description of ~set~ -command and its parameters:

#+RESULTS: help-set
#+begin_example
set - Configure

Configure voltage and current setpoints on DPS. Does NOT change
output status.

     U_set  : Voltage to set
     I_set  : Current limit to set

Notice:
- parameters above can be in any order
- parameters are optional and they MAY be left out
#+end_example


Command ~??~ gives more detailed help on a command 

#+BEGIN_SRC bash :eval no-export :results output :noweb yes :exports results
 ebDps ?? command='??'
#+END_SRC

#+RESULTS:
: ?? - List command parameters
: 
:    command  : Command to give help on (None: help on main menu)
: 
: Notice:
: - parameters above can be in any order
: - parameters are optional and they MAY be left out



* Commands and parameters

# ebenchDps/ebenchDps_main.py
#+name: ebDps
#+BEGIN_SRC python :eval no :results output :noweb no :session *Python* :exports none
ebDps
#+END_SRC

** =status= - Read DSP status

#+BEGIN_SRC bash :eval no-export :results output :noweb yes :exports results
ebDps ?? command=status
#+END_SRC

#+RESULTS:
#+begin_example
status - Read DSP status

Return DPS status as a JSON document with fields:

- U_set: Voltage setpoint
- I_set: Current setpoint
- U_out: Output voltage display value
- I_out: Output current display value
- P_out: Output power  display value
- U_in: input voltage
- onoff: Output on/off
- lock: Key lock
- protect: Protection status
- cv/cc: Constant Voltage / Constant Current status
- on/off: switch output state
- b_led: backligh brighness level
- model: product model
- version: firmware version

,*No parameters*

Notice:
- parameters above can be in any order
- parameters are optional and they MAY be left out
#+end_example

Example:

#+BEGIN_SRC bash :eval no-export :results output :noweb yes :exports both
ebDps status
#+END_SRC

#+RESULTS:
#+begin_example
{'I_out': 0.0,
 'I_set': 2.01,
 'P_out': 0.0,
 'U_in': 20.32,
 'U_out': 0.0,
 'U_set': 5.0,
 'b_led': 4,
 'cv/cc': 0,
 'lock': 0,
 'model': 5020,
 'onoff': 0,
 'protect': 0,
 'version': 1.6}
#+end_example



** =off= - Turn off

#+BEGIN_SRC bash :eval no-export :results output :noweb yes :exports results
ebDps ?? command=off
#+END_SRC

#+RESULTS:
#+begin_example
off - Turn off

Turn output off
        

,*No parameters*

Notice:
- parameters MUST be given in the order listed above
- parameters are optional and they MAY be left out
#+end_example

For example, use =off= command followed by =status= -command to
reading device status and ensure that field ~'onoff': 0,~.

#+BEGIN_SRC bash :eval no-export :results output :noweb yes :exports both
ebDps off status
#+END_SRC

#+RESULTS:
#+begin_example
{'I_out': 0.0,
 'I_set': 1.4,
 'P_out': 0.0,
 'U_in': 10.38,
 'U_out': 0.0,
 'U_set': 6.0,
 'b_led': 4,
 'cv/cc': 0,
 'lock': 0,
 'model': 5020,
 'onoff': 0,
 'protect': 0,
 'version': 1.6}
#+end_example

In the output notice ='onoff': 0=


** =set= -configure

#+BEGIN_SRC bash :eval no-export :results output :noweb yes :exports results
ebDps ?? command=set
#+END_SRC

#+RESULTS:
#+begin_example
set - Configure

Configure voltage and current setpoints on DPS. Does NOT change
output status.

     U_set  : Voltage to set
     I_set  : Current limit to set

Notice:
- parameters MUST be given in the order listed above
- parameters are optional and they MAY be left out
#+end_example


The example below, sets output voltaget to =3.3= volts and current
limit to =0.1= ampere, and switches the output on. Status read shows
values =U_set= and =U_out=, and =onoff: 1=


#+BEGIN_SRC bash :eval no-export :results output :noweb yes :exports both
ebDps off set U_set=3.3 I_set=0.1 status
#+END_SRC

#+RESULTS:
#+begin_example
{'I_out': 0.0,
 'I_set': 0.1,
 'P_out': 0.0,
 'U_in': 24.04,
 'U_out': 0.0,
 'U_set': 3.3,
 'b_led': 4,
 'cv/cc': 0,
 'lock': 0,
 'model': 5020,
 'onoff': 0,
 'protect': 0,
 'version': 1.6}
#+end_example

In the output above, notice:
- ~'onoff': 0~: device output is off 
- ~U_out, I_out, P_out=0~: no output 
- ~U_set': 3.3~: voltage setpoint configured
- ~I_set': 0.1~: currrent limit configured


** =on= turn on

#+BEGIN_SRC bash :eval no-export :results output :noweb yes :exports results
ebDps ?? command=on
#+END_SRC

#+RESULTS:
#+begin_example
on - Turn on

Turn output on
        

,*No parameters*

Notice:
- parameters above can be in any order
- parameters are optional and they MAY be left out
#+end_example

The example configures setpoint 5V,0.2A (command =set= and with
parameters =U_set=, ~I_set~) , turns the DPS output on (command =on=),
and read device status (command =status=)

#+BEGIN_SRC bash :eval no-export :results output :noweb yes :exports results
ebDps set U_set=5 I_set=0.2 on status
#+END_SRC



In the output above notice:
- 'U_set': 5.0,
-  I_set': 0.2,
-  'onoff': 1,


** =csv= - CSV driver

#+BEGIN_SRC bash :eval no-export :results output :noweb yes :exports results
ebDps ?? command=csv
#+END_SRC

#+RESULTS:
#+begin_example
csv - CSV driver

Control devices using csvfile (default delimiter comma=,)

Csv file header:

- step_time:
- voltage:
- current: 

Example CSV content


step_time,voltage,current
2,0.5,0.1
3,2.5,0.2
3,3.3,0.3

  filePath  : Path to CSV -file
        on  : Switch DPS on

Notice:
- parameters MUST be given in the order listed above
- parameters are optional and they MAY be left out
#+end_example


Using CSV file ~example.csv~

#+BEGIN_SRC csv :tangle example.csv :exports code
step_time,voltage,current
2,0.5,0.1
3,2.5,0.2
3,3.3,0.3
#+END_SRC

#+BEGIN_SRC bash :eval no-export :results output :noweb yes
ebDps csv filePath=example.csv on=yes
#+END_SRC

#+RESULTS:
: ...



** =scan_ports= - list tty devices

#+BEGIN_SRC bash :eval no-export :results output :noweb yes :exports results
ebDps ?? command=scan_ports
#+END_SRC

#+RESULTS:
: scan_ports - Scan for device serial address
: 
: Lists serial port names
: 
: *No parameters*
: 
: Notice:
: - parameters above can be in any order
: - parameters are optional and they MAY be left out


Example output on development environment

#+BEGIN_SRC bash :eval no-export :results output :noweb yes
ebDps scan_ports
#+END_SRC

#+RESULTS:
: ports=['/dev/ttyUSB0', '/dev/ttyprintk', '/dev/ttyS31', '/dev/ttyS30', '/dev/ttyS29', '/dev/ttyS28', '/dev/ttyS27', '/dev/ttyS26', '/dev/ttyS25', '/dev/ttyS24', '/dev/ttyS23', '/dev/ttyS22', '/dev/ttyS21', '/dev/ttyS20', '/dev/ttyS19', '/dev/ttyS18', '/dev/ttyS17', '/dev/ttyS16', '/dev/ttyS15', '/dev/ttyS14', '/dev/ttyS13', '/dev/ttyS12', '/dev/ttyS11', '/dev/ttyS10', '/dev/ttyS9', '/dev/ttyS8', '/dev/ttyS7', '/dev/ttyS6', '/dev/ttyS5', '/dev/ttyS4', '/dev/ttyS3', '/dev/ttyS2', '/dev/ttyS1', '/dev/ttyS0']
: ['/dev/ttyUSB0']




** =_ttys= - Check serial device

#+BEGIN_SRC bash :eval no-export :results output :noweb yes :exports both
ebDps ?? command=_ttys
#+END_SRC

#+RESULTS:
: _ttys - List serial devices
: 
: List tty devices found
: 
:          i  : Index to to show
: 
: Notice:
: - parameters above can be in any order
: - parameters are optional and they MAY be left out


#+BEGIN_SRC bash :eval no-export :results output :noweb yes :exports both
ebDps _ttys
#+END_SRC

#+RESULTS:
: /dev/ttyUSB0

Give index value to choose one of the ttys found

#+BEGIN_SRC bash :eval no-export :results output :noweb yes :exports both
ebDps _ttys i=0
#+END_SRC

#+RESULTS:
: /dev/ttyUSB0

Using index values out of range results to an error

#+BEGIN_SRC bash :eval no-export :results output :noweb yes :exports both
(ebDps _ttys i=10 2>&1 || true ) | grep IndexError
#+END_SRC

#+RESULTS:
: IndexError: list index out of range




** =_info= -command

#+BEGIN_SRC bash :eval no-export :results output :noweb yes :exports both
ebDps _info
#+END_SRC

#+RESULTS:
: {'dpsModel': 5020,
:  'dpsVersion': 1.6,
:  'ebenchDps': '0.0.5-SNAPSHOT, (ebench.version:0.0.12-pre7)'}


* Document information

** Version used

#+BEGIN_SRC bash :eval no-export :results output :noweb yes :exports results
ebDps _version
#+END_SRC

#+RESULTS:
: 0.0.5-pre3, (ebench.version:0.0.12-pre7)




* Notes                                                            :noexport:
** Documents 

[[file:~/work/data-sheets/kiina/dps5020]]

COmmunication protocol [[file:~/work/data-sheets/kiina/dps5020/DPS5020%20CNC%20Communication%20Protocol%20V1.2.pdf]]
Modbus spec: https://modbus.org/docs/PI_MBUS_300.pdf




* Fin                                                              :noexport:

   #+RESULTS:

   # Local Variables:
   # org-confirm-babel-evaluate: nil
   # org-make-toc-mode: t
   # End:




ChangeLog
2022-03-13  jj  <jj@santtu>

	* ebenchDps/ebenchDps.py: dps8020.ini default file from current path
	(ModbusInstrument.initInstrument): lazy initInstrument
	(DPSInstrument.serial_ports): added
	(run): CMD_SCAN added



serialports.py
import serial
import glob

def serial_ports():
	""" Lists serial port names
		:raises EnvironmentError:
			On unsupported or unknown platforms
		:returns:
			A list of the serial ports available on the system
	"""
	if sys.platform.startswith('win'):
		ports = ['COM%s' % (i + 1) for i in range(256)]
	elif sys.platform.startswith('linux') or sys.platform.startswith('cygwin'):
		# this excludes your current terminal "/dev/tty"
		ports = glob.glob('/dev/tty[A-Za-z]*')
		print( "ports={}".format(ports))                     
	elif sys.platform.startswith('darwin'):
		ports = glob.glob('/dev/tty.*')
	else:
		raise EnvironmentError('Unsupported platform')

	result = []
	for port in ports:
		try:
			s = serial.Serial(port)
			s.close()
			result.append(port)
		except (OSError, serial.SerialException):
			pass
	return result



ebenchDps/ebenchDps_main.py
#!/usr/bin/env python3


# main for instrument controller define in module
from .ebenchDps import run
# from ebenchDps import run

from absl import app, flags, logging
from absl.flags import FLAGS

# Run time configurations of instrument controller
flags.DEFINE_string('port', "/dev/ttyUSB0", "MODBUS device port")


def _main( _argv ):
    logging.set_verbosity(FLAGS.debug)
    logging.info( "FLAGS.port={}".format(FLAGS.port))
    menuController = run(
        _argv
        , port=FLAGS.port          # pass run time configuration parameters to controller
        , captureDir=FLAGS.captureDir
        , recordingDir=FLAGS.recordingDir
        , outputTemplate=FLAGS.outputTemplate 
    )
    menuController.close()


def main():
    try:
        app.run(_main)
    except SystemExit:
        pass
    
    
if __name__ == '__main__':
    main()


ebenchDps/ebenchDps.py
from ebench import Instrument
from ebench import MenuCtrl

from ebench import usage, usageCommand, menuStartRecording, menuStopRecording, menuScreenShot

from ebench import version as ebenchVersion

# wraps https://forum-raspberrypi.de/forum/thread/46324-ansteuerung-programmierbarer-stepdown-regler-dps/
from .dps_modbus import serial_ports, Serial_modbus, Import_limits, Dps5005
from .CMDS import CMD_DPS
import os
import sys
# from time import sleep
from absl import logging
import csv
#from functools import reduce
import pathlib
import errno
import glob
import serial

# ------------------------------------------------------------------
# Usage 
CMD="ebenchDps"

SYNOPSIS="Control DPS8020 Digital Power Supply"

USAGE_TEXT = """

The tool :

- query Digital Power supply status e.g. voltage/current setpoint
- configure voltage and current setpoint
- turn output on and off
- use CSV file to generate output voltage output

"""

# ------------------------------------------------------------------
# delegation pattern

# https://gist.github.com/kkew3/2d61463397a77a898be12ac766c31c12
def delegates(method_names, to):
    def dm(method_name):
        def fwraps(self, *args, **kwargs):
            wrappedf = getattr(getattr(self, to), method_name)
            return wrappedf(*args, **kwargs)
        fwraps.__name__ = method_name
        return fwraps
    def cwraps(cls):
        for name in method_names:
            setattr(cls, name, dm(name))
        return cls
    return cwraps


@delegates( ["read", "read_block", "write", "write_block", ], to="ser")
class ModbusInstrument(Instrument):
    """Abstract class for an instrument, which can be controlled using
    modbus interface.

    Wraps 'dps_modbus.Serial_modbus' -class in property 'ser'

    """

    def __init__(self, port, addr=1, baud_rate=9600, byte_size=8):
        # Save contructor values for lazy initInstrument
        self.port = port
        self.addr = addr
        self.baud_rate = baud_rate
        self.byte_size = byte_size
        # self.initInstrument()
        # self.ser = Serial_modbus( port1=port, addr=addr, baud_rate=baud_rate, byte_size=byte_size)

    def initInstrument(self):
        """Construct 'Serial_modbus' object using  self.properties
        
        Use case: lazy initializer

        """
        logging.info( "initInstrument: Serial_modbus constructed port=f{self.port}, addr={self.addr},  baud_rate={self.baud_rate}, byte_size={self.byte_size}" )
        return Serial_modbus( port1=self.port, addr=self.addr, baud_rate=self.baud_rate, byte_size=self.byte_size)
        
    # .................................
    # properties
    @property
    def addr(self) -> int :
        if not hasattr(self, "_addr"):
             return None
        return self._addr

    @addr.setter
    def addr( self, addr:int):
        self._addr = addr

    @property
    def port(self) -> str :
        if not hasattr(self, "_port"):
             return None
        return self._port

    @port.setter
    def port( self, port:str):
        self._port = port

    @property
    def baud_rate(self) -> int :
        if not hasattr(self, "_baud_rate"):
             return None
        return self._baud_rate

    @baud_rate.setter
    def baud_rate( self, baud_rate:int):
        self._baud_rate = baud_rate

    @property
    def byte_size(self) -> int :
        if not hasattr(self, "_byte_size"):
             return None
        return self._byte_size

    @byte_size.setter
    def byte_size( self, byte_size:int):
        self._byte_size = byte_size


    @property
    def ser(self) -> Serial_modbus :
        """Lazy constructor"""
        if not hasattr(self, "_ser"):
            self._ser = self.initInstrument()
        return self._ser

    @ser.setter
    def ser( self, ser:Serial_modbus):
        self._ser = ser

@delegates( ["read_all"
             , "function"
             , "functions"
             , "voltage_in"
             , "voltage_set"
             , "current_set"
             , "delay"
             # , "voltage"
             # , "current"
             , "version"
             # , "lock"
             # , "protect"
             , "model"
             # , "cv_cc"
             , "onoff"], to="dps") # using dps_modbus.Dps5005 class methods
class DPSApi(ModbusInstrument):
    """
    Instrument API services
    """

    def __init__(self, port:str, addr:int=1, propertiesPath = None ):
        if propertiesPath is None:
            propertiesPath = os.path.join( pathlib.Path(__file__).parent.resolve(), "dps5020.ini")
        logging.info( f"DPSApi.__init__: propertiesPath={propertiesPath}")
        if not os.path.exists( propertiesPath ):
            raise FileNotFoundError( errno.ENOENT, os.strerror(errno.ENOENT), propertiesPath )
        super().__init__(port=port, addr=addr)
        self.limits = Import_limits( propertiesPath )
        self.dps = Dps5005(self, self.limits)

    # .................................
    # properties

    @property
    def limits(self) -> Import_limits :
        if not hasattr(self, "_limits"):
             return None
        return self._limits

    @limits.setter
    def limits( self, limits:Import_limits):
        self._limits = limits

    @property
    def dps(self) -> Dps5005 :
        if not hasattr(self, "_dps"):
             return None
        return self._dps

    @dps.setter
    def dps( self, dps:Dps5005):
        self._dps = dps


class DPSInstrument(DPSApi):
    """
    Facade presented to user.

    """
    
    CSV_COLUMNS=["step_time", "voltage", "current"]
    
    def __init__(self, port:str):
        super().__init__(port=port)

    # ------------------------------------------------------------------
    # Genral utilities
        

    def isTrueOrFalse( self,value:str )-> bool:
        if value is None or not value: return False
        if value.upper() in [ "T", "Y", "TRUE", "ON","YES", "1", "K" ]: return True
        return False

    
    def csv2dict( self, filePath:str, delimiter=",") -> dict:
        """Slurp CSV -file in 'filePath' into dictionary mapping header
        columns to data-arrays

        :return: Dict[str,List]

        """
        with open( filePath, mode="r") as csv_file:
            csv_reader = csv.DictReader(csv_file, delimiter=delimiter)
            line_count = 0
            csvDict = None
            for row in csv_reader:
                logging.debug( "csv2dict: row={}".format(row) )
                if line_count == 0:
                    csvDict = { k: [v] for k,v in row.items()}
                else:
                    for i,key in enumerate(row.keys()):
                        csvDict[key].append(row[key])
                line_count += 1
            return csvDict

    def spinner(self,value="."):
        """Output spinner 'value' indicating progress 

        """
        print( value, end="")
        sys.stdout.flush()


    # ------------------------------------------------------------------
    # User facade
    def lese(self):
        """
        Return DPS status as a JSON document with fields:

        - U_set: Voltage setpoint
        - I_set: Current setpoint
        - U_out: Output voltage display value
        - I_out: Output current display value
        - P_out: Output power  display value
        - U_in: input voltage
        - onoff: Output on/off
        - lock: Key lock
        - protect: Protection status
        - cv/cc: Constant Voltage / Constant Current status
        - on/off: switch output state
        - b_led: backligh brighness level
        - model: product model
        - version: firmware version
        
        """
        readArr = self.read_all()
        # ref. API https://www.mediafire.com/file/tw96e28oggw7dai/DPS5020_CNC_Communication__Protocol_V1.2.pdf
        # Register Map for this device
        retJson = {
            "U_set": readArr[0],
            "I_set": readArr[1],
            "U_out": readArr[2],
            "I_out": readArr[3],
            "P_out": readArr[4],
            "U_in": readArr[5],
            "lock": readArr[6],
            "protect": readArr[7],
            "cv/cc": readArr[8],
            "onoff": readArr[9],
            "b_led": readArr[10],
            "model": readArr[11],
            "version": readArr[12],
        }
        return retJson


    def set(self, U_set=None, I_set=None ):
        """Configure voltage and current setpoints on DPS. Does NOT change
         output status.

        """
        logging.info( "set: U_set={}, I_set={}".format(U_set, I_set) )
        if not not U_set and U_set is not None:
            self.voltage_set( RWaction="w", value=float(U_set))
        if not not I_set and I_set is not None:
            self.current_set( RWaction="w", value=float(I_set))
            
    def on(self):
        """Turn output on
        """
        self.onoff(RWaction="w", value=1)
        
    def off(self):
        """Turn output off
        """
        self.onoff(RWaction="w", value=0)


    def info(self):
        """
        Returns JSON document with following fields
        
        - ebenchDps: ebenchDps application version string
        - dpsModel: DPS device model number
        - dpsVersion: DPS version number
        
        """
        jsonRet = {
            "ebenchDps": version(),
            "dpsModel": self.model(),
            "dpsVersion": self.version(),
        }
        return jsonRet


    def serial_ports( self):
        """
            Lists serial port names
        """
        # Lists serial port names
        #        :raises EnvironmentError:
        #        	On unsupported or unknown platforms
        #        :returns:
        #        	A list of the serial ports available on the system
        if sys.platform.startswith('win'):
            ports = ['COM%s' % (i + 1) for i in range(256)]
        elif sys.platform.startswith('linux') or sys.platform.startswith('cygwin'):
            # this excludes your current terminal "/dev/tty"
            ports = glob.glob('/dev/tty[A-Za-z]*')
            print( "ports={}".format(ports))                     
        elif sys.platform.startswith('darwin'):
            ports = glob.glob('/dev/tty.*')
        else:
            raise EnvironmentError('Unsupported platform')

        result = []
        for port in ports:
                try:
                        s = serial.Serial(port)
                        s.close()
                        result.append(port)
                except (OSError, serial.SerialException):
                        pass
        return result


    def csv(self, filePath:str, on:str=None, delimiter:str=","):
        """
        Control devices using csvfile (default delimiter comma=,)

        Csv file header:

        - step_time:
        - voltage:
        - current: 

        Example CSV content


        step_time,voltage,current
        2,0.5,0.1
        3,2.5,0.2
        3,3.3,0.3

        
        """
        logging.info( "csv: filePath={}".format(filePath))

        # Read filePath into dictionary
        csvDict = self.csv2dict(filePath=filePath, delimiter=",")

        # Validate data
        if csvDict is None:
            raise ValueError( "Could not read csv from filePath={}".format(filePath))
        for col in DPSInstrument.CSV_COLUMNS:
            if col not in csvDict:
                raise ValueError( "Missing column {}, expect to have columns {} in CSV file {}".format(col, DPSInstrument.CSV_COLUMNS, csvDict.keys()))

            
        # Init, Start times && Looop
        self.delay(0.0)
        self.voltage_set( "w", 0.0)
        self.current_set( "w", 0.0)

        # Turn on - unless dry run
        if self.isTrueOrFalse(on): self.on()
        
        for i,step_time in enumerate(csvDict["step_time"]):
            # We are advancing
            self.spinner()
            # One step
            self.voltage_set( "w", float(csvDict["voltage"][i]) )
            self.current_set( "w", float(csvDict["current"][i]) )
            self.delay(float(step_time))

        # No output
        return None
    
# ------------------------------------------------------------------
# Module services
        
def dpsVersion():
    versionPath = os.path.join( os.path.dirname( __file__), "VERSION")
    with open( versionPath, "r") as fh:
        version = fh.read().rstrip()
    return version

def version():
    return( "{}, (ebench.version:{})".format(dpsVersion(),  ebenchVersion() ))

def ttyShow(i=None):
    """List tty devices found"""
    ttys = serial_ports()
    if not not i and i is not None:
        return ttys[int(i)]
    else:
        return ",".join(ttys)
    
# ------------------------------------------------------------------
# Menu
CMD_STATUS="status"
CMD_SET="set"
CMD_ON="on"
CMD_OFF="off"
CMD_CSV="csv"
CMD_SCAN="scan_ports"

# Hidden commands
CMD_SHOW_TTYS="_ttys"
CMD_INFO="_info"

ttyShowPar = {
    'i':"Index to to show"
}
setPar = {
    'U_set':"Voltage to set",
    'I_set':"Current limit to set",
}

csvPar = {
    'filePath':"Path to CSV -file",
    'on':"Switch DPS on"
}


# Initial values for menu command parameters
defaults = {
    CMD_SET : {
        "U_set": 0.0,     # default && remember
        "I_set": 0.0,     # remember last value
    },
    CMD_CSV : {
        "on": "NO",       # default is not switch on
    },
}
    
# ------------------------------------------------------------------
# Bind instrument controller classes to ebench toolset
def run( _argv
         , port:str
         , runMenu:bool = True
         , outputTemplate=None, captureDir=None, recordingDir=None ):
    """Construct DPSIntrument for controlling DPS Digital Power Supply

    :port: device address of DPS mobbus interface

    :runMenu: default True, standalone application call REPL-loop
    'menuController.mainMenu()', subMenu constructs 'menuController'
    without executing the loop

    :outputTemplate: if None(default): execute cmds/args, else (not
    None): map menu actions to strings using 'outputTemplate'

    :recordingDir: directory where interactive session recordings are
    saved to (defaults to 'FLAGS.recordingDir')

    :captureDir: directory where screenshots are made, defaults to
    'FLAGS.captureDir'

    :return: MenuCtrl (wrapping instrument)

    """

    # 'instrument' controlled by application
    logging.info( "run: contstruct DPSInstrument, port={}".format(port))
    instrument = DPSInstrument(port=port) 

    # Wrap instrument with 'MenuCtrl'
    menuController = MenuCtrl( args=_argv,instrument=instrument
                             , prompt="[q=quit,?=commands,??=help on command]"
                             , outputTemplate=outputTemplate )

    mainMenu = {
        CMD                      : MenuCtrl.MENU_SEPATOR_TUPLE,
        # Application menu 
        # CMD_GREET                : ( "Say hello", greetPar, instrument.sayHello ),
        CMD_STATUS               : ("Read DSP status", None, instrument.lese ),
        CMD_SET                  : ("Configure", setPar, instrument.set ),
        CMD_ON                   : ("Turn on", None, instrument.on ),
        CMD_CSV                  : ("CSV driver", csvPar, instrument.csv ),        
        CMD_OFF                  : ("Turn off", None, instrument.off ),
        
        "Management"             : MenuCtrl.MENU_SEPATOR_TUPLE,
        CMD_SCAN                 : ("Scan for device serial address", None, instrument.serial_ports),

        "Util"                   : MenuCtrl.MENU_SEPATOR_TUPLE,
        MenuCtrl.MENU_REC_START  : ( "Start recording", None, menuStartRecording(menuController) ),
        MenuCtrl.MENU_REC_SAVE   : ( "Stop recording", MenuCtrl.MENU_REC_SAVE_PARAM, menuStopRecording(menuController, recordingDir=recordingDir) ),
        MenuCtrl.MENU_SCREEN     : ( "Take screenshot", MenuCtrl.MENU_SCREENSHOT_PARAM,
                                     menuScreenShot(instrument=instrument,captureDir=captureDir,prefix="Capture-" )),
        MenuCtrl.MENU_HELP       : ( "List commands", None,
                                    lambda **argV: usage(cmd=CMD_DPS, mainMenu=mainMenu, synopsis=SYNOPSIS, usageText=USAGE_TEXT)),
        MenuCtrl.MENU_HELP_CMD   : ( "List command parameters", MenuCtrl.MENU_HELP_CMD_PARAM,
                                 lambda **argV: usageCommand(mainMenu=mainMenu, **argV )),

        "Quit"                   : MenuCtrl.MENU_SEPATOR_TUPLE,
        MenuCtrl.MENU_QUIT       : MenuCtrl.MENU_QUIT_TUPLE,

        # Hidden commands
        MenuCtrl.MENU_VERSION    : ( "Output version number", None, version ),
        CMD_SHOW_TTYS            : ( "List serial devices", ttyShowPar, ttyShow ),
        CMD_INFO                 : ( "Application info", None, instrument.info ),
    }

    menuController.setMenu( menu = mainMenu, defaults = defaults)

    # Interactive use starts REPL-loop
    if runMenu: menuController.mainMenu()

    # menuController.close() call after returning from run()
    return menuController
